File: ios/BatteryHolder/DesignSystem/Theme.swift
import SwiftUI

// MARK: - Hex color helpers

extension Color {
    init(hex: UInt32, alpha: Double = 1) {
        let r = Double((hex >> 16) & 0xFF) / 255
        let g = Double((hex >> 8) & 0xFF) / 255
        let b = Double(hex & 0xFF) / 255
        self.init(.sRGB, red: r, green: g, blue: b, opacity: alpha)
    }
}

/// Resolves to `light` or `dark` at runtime based on the trait environment.
private func adaptive(_ light: UInt32, _ dark: UInt32) -> Color {
    Color(uiColor: UIColor { traits in
        traits.userInterfaceStyle == .dark
            ? UIColor(Color(hex: dark))
            : UIColor(Color(hex: light))
    })
}

// MARK: - Theme namespace

/// Design-token entry point. See docs/DESIGN_TOKENS.md for the reference table.
enum Theme {
    static let color = AppColor()
    static let spacing = Spacing()
    static let radius = Radius()
    static let font = Typography()
    static let elevation = Elevation()
}

// MARK: - Color tokens

struct AppColor {
    // Brand & accent
    let brand         = adaptive(0x0A84FF, 0x0A84FF)
    let brandPressed  = adaptive(0x0060DF, 0x409CFF)
    let accent        = adaptive(0x30D158, 0x30D158)

    // Battery status scale
    let batteryGood     = adaptive(0x30D158, 0x32D74B)
    let batteryMedium   = adaptive(0xFF9F0A, 0xFFB340)
    let batteryLow      = adaptive(0xFF9500, 0xFF9F0A)
    let batteryCritical = adaptive(0xFF453B, 0xFF6961)

    // Feedback
    let success = adaptive(0x248A3D, 0x30D158)
    let warning = adaptive(0xB25000, 0xFF9F0A)
    let danger  = adaptive(0xD70015, 0xFF453B)

    // Surfaces & text
    let background      = adaptive(0xF2F2F7, 0x000000)
    let surface         = adaptive(0xFFFFFF, 0x1C1C1E)
    let surfaceElevated = adaptive(0xFFFFFF, 0x2C2C2E)
    let border          = adaptive(0xE5E5EA, 0x38383A)
    let textPrimary     = adaptive(0x1C1C1E, 0xFFFFFF)
    let textSecondary   = adaptive(0x6C6C70, 0x98989F)
    let textOnBrand     = adaptive(0xFFFFFF, 0xFFFFFF)

    /// Maps a battery percentage (0...1) to its status color.
    func battery(forPercentage pct: Double) -> Color {
        switch pct {
        case ..<0.10: return batteryCritical
        case ..<0.25: return batteryLow
        case ..<0.60: return batteryMedium
        default:      return batteryGood
        }
    }
}

// MARK: - Spacing (4pt grid)

struct Spacing {
    let xxs: CGFloat = 2
    let xs: CGFloat = 4
    let sm: CGFloat = 8
    let md: CGFloat = 12
    let lg: CGFloat = 16
    let xl: CGFloat = 24
    let xxl: CGFloat = 32
    let xxxl: CGFloat = 48
}

// MARK: - Radius

struct Radius {
    let sm: CGFloat = 8
    let md: CGFloat = 12
    let lg: CGFloat = 16
    let xl: CGFloat = 24
    let pill: CGFloat = 999
}

// MARK: - Typography

struct Typography {
    let largeTitle  = Font.system(size: 34, weight: .bold)
    let title       = Font.system(size: 22, weight: .semibold)
    let headline    = Font.system(size: 17, weight: .semibold)
    let body        = Font.system(size: 17, weight: .regular)
    let callout     = Font.system(size: 16, weight: .regular)
    let subheadline = Font.system(size: 15, weight: .regular)
    let footnote    = Font.system(size: 13, weight: .regular)
    let caption     = Font.system(size: 12, weight: .regular)
    let mono        = Font.system(size: 17, weight: .regular, design: .monospaced)
    let monoLarge   = Font.system(size: 40, weight: .semibold, design: .monospaced)
}

// MARK: - Elevation

struct Elevation {
    struct Shadow { let color: Color; let radius: CGFloat; let y: CGFloat }
    let card    = Shadow(color: .black.opacity(0.08), radius: 8,  y: 2)
    let raised  = Shadow(color: .black.opacity(0.12), radius: 16, y: 6)
    let overlay = Shadow(color: .black.opacity(0.18), radius: 32, y: 12)
}

File: android_app/lib/design_system/theme.dart
import 'package:flutter/material.dart';

/// Design-token entry point, a 1:1 port of
/// `ios/BatteryHolder/DesignSystem/Theme.swift`.
/// See docs/DESIGN_TOKENS.md for the reference table.
///
/// Colors are *adaptive*: every token carries a light and a dark value and
/// resolves against the ambient [Brightness], the same way the iOS dynamic
/// `UIColor` does. Views read them through `Theme.colorOf(context)`.
class AppTheme {
  static const spacing = Spacing();
  static const radius = Radius();
  static const font = Typography();
  static const elevation = Elevation();

  /// Resolve the color set for the current brightness.
  static AppColor colorOf(BuildContext context) =>
      AppColor(Theme.of(context).brightness == Brightness.dark);

  static const AppColor light = AppColor(false);
  static const AppColor dark = AppColor(true);
}

Color _hex(int hex, [double opacity = 1]) => Color(hex | 0xFF000000)
    .withValues(alpha: opacity);

// MARK: - Color tokens

class AppColor {
  /// True when resolving against a dark trait environment.
  final bool isDark;

  const AppColor(this.isDark);

  Color _adaptive(int light, int dark) => _hex(isDark ? dark : light);

  // Brand & accent
  Color get brand => _adaptive(0x0A84FF, 0x0A84FF);
  Color get brandPressed => _adaptive(0x0060DF, 0x409CFF);
  Color get accent => _adaptive(0x30D158, 0x30D158);

  // Battery status scale
  Color get batteryGood => _adaptive(0x30D158, 0x32D74B);
  Color get batteryMedium => _adaptive(0xFF9F0A, 0xFFB340);
  Color get batteryLow => _adaptive(0xFF9500, 0xFF9F0A);
  Color get batteryCritical => _adaptive(0xFF453B, 0xFF6961);

  // Feedback
  Color get success => _adaptive(0x248A3D, 0x30D158);
  Color get warning => _adaptive(0xB25000, 0xFF9F0A);
  Color get danger => _adaptive(0xD70015, 0xFF453B);

  // Surfaces & text
  Color get background => _adaptive(0xF2F2F7, 0x000000);
  Color get surface => _adaptive(0xFFFFFF, 0x1C1C1E);
  Color get surfaceElevated => _adaptive(0xFFFFFF, 0x2C2C2E);
  Color get border => _adaptive(0xE5E5EA, 0x38383A);
  Color get textPrimary => _adaptive(0x1C1C1E, 0xFFFFFF);
  Color get textSecondary => _adaptive(0x6C6C70, 0x98989F);
  Color get textOnBrand => _adaptive(0xFFFFFF, 0xFFFFFF);

  /// Maps a battery percentage (0...1) to its status color.
  Color battery(double pct) {
    if (pct < 0.10) return batteryCritical;
    if (pct < 0.25) return batteryLow;
    if (pct < 0.60) return batteryMedium;
    return batteryGood;
  }
}

// MARK: - Spacing (4pt grid)

class Spacing {
  const Spacing();
  final double xxs = 2;
  final double xs = 4;
  final double sm = 8;
  final double md = 12;
  final double lg = 16;
  final double xl = 24;
  final double xxl = 32;
  final double xxxl = 48;
}

// MARK: - Radius

class Radius {
  const Radius();
  final double sm = 8;
  final double md = 12;
  final double lg = 16;
  final double xl = 24;
  final double pill = 999;
}

// MARK: - Typography
//
// iOS uses SF Pro; Android's platform face is Roboto. Sizes and weights match
// the Swift tokens exactly so the vertical rhythm is identical.

class Typography {
  const Typography();

  TextStyle get largeTitle =>
      const TextStyle(fontSize: 34, fontWeight: FontWeight.w700);
  TextStyle get title =>
      const TextStyle(fontSize: 22, fontWeight: FontWeight.w600);
  TextStyle get headline =>
      const TextStyle(fontSize: 17, fontWeight: FontWeight.w600);
  TextStyle get body =>
      const TextStyle(fontSize: 17, fontWeight: FontWeight.w400);
  TextStyle get callout =>
      const TextStyle(fontSize: 16, fontWeight: FontWeight.w400);
  TextStyle get subheadline =>
      const TextStyle(fontSize: 15, fontWeight: FontWeight.w400);
  TextStyle get footnote =>
      const TextStyle(fontSize: 13, fontWeight: FontWeight.w400);
  TextStyle get caption =>
      const TextStyle(fontSize: 12, fontWeight: FontWeight.w400);

  /// `RobotoMono` ships with Flutter's Material assets and is the closest
  /// match to SF Mono for the ADC / voltage readouts.
  TextStyle get mono => const TextStyle(
      fontSize: 17, fontWeight: FontWeight.w400, fontFamily: 'monospace');
  TextStyle get monoLarge => const TextStyle(
      fontSize: 40, fontWeight: FontWeight.w600, fontFamily: 'monospace');
}

// MARK: - Elevation

class Elevation {
  const Elevation();

  List<BoxShadow> get card => const [
        BoxShadow(color: Color(0x14000000), blurRadius: 8, offset: Offset(0, 2)),
      ];
  List<BoxShadow> get raised => const [
        BoxShadow(
            color: Color(0x1F000000), blurRadius: 16, offset: Offset(0, 6)),
      ];
  List<BoxShadow> get overlay => const [
        BoxShadow(
            color: Color(0x2E000000), blurRadius: 32, offset: Offset(0, 12)),
      ];
}

// MARK: - Motion

class Motion {
  static const quick = Duration(milliseconds: 150);
  static const standard = Duration(milliseconds: 350);
  static const emphasis = Duration(milliseconds: 500);
}

// MARK: - MaterialTheme wiring

/// Builds the [ThemeData] the app runs on, seeded from the tokens above so
/// stock Material widgets (dialogs, pickers, the tab bar) inherit the same
/// palette as the hand-built components.
ThemeData buildAppTheme(Brightness brightness) {
  final c = AppColor(brightness == Brightness.dark);
  final base = brightness == Brightness.dark ? ThemeData.dark() : ThemeData.light();

  return base.copyWith(
    brightness: brightness,
    scaffoldBackgroundColor: c.background,
    canvasColor: c.background,
    dividerColor: c.border,
    colorScheme: ColorScheme.fromSeed(
      seedColor: c.brand,
      brightness: brightness,
    ).copyWith(
      primary: c.brand,
      onPrimary: c.textOnBrand,
      surface: c.surface,
      onSurface: c.textPrimary,
      error: c.danger,
    ),
    appBarTheme: AppBarTheme(
      backgroundColor: c.background,
      foregroundColor: c.textPrimary,
      surfaceTintColor: Colors.transparent,
      elevation: 0,
      scrolledUnderElevation: 0,
      centerTitle: false,
      titleTextStyle: TextStyle(
        // Matches SwiftUI's inline `.navigationTitle` weight/size.
        fontSize: 34,
        fontWeight: FontWeight.w700,
        color: c.textPrimary,
      ),
    ),
    dividerTheme: DividerThemeData(color: c.border, thickness: 0.5, space: 1),
    progressIndicatorTheme: ProgressIndicatorThemeData(
      color: c.brand,
      linearTrackColor: c.border,
    ),
    textSelectionTheme: TextSelectionThemeData(cursorColor: c.brand),
  );
}
